File: 2018/solutions/prob05.cpp
#include <iostream>
#include <string>

std::string foldword(const std::string &line);

int main() {
  std::string word;

  while (std::cin >> word) {
    std::cout << foldword(word) << std::endl;
  }

  return EXIT_SUCCESS;
}

std::string foldword(const std::string &line)
{
  bool state = true;
  size_t lineLength = line.size();
  size_t k = 0, i = 0, j = lineLength - 1;
  std::string reordered(lineLength, '\0');

  while (i <= j) {
    if (state) {
      reordered[k] = line[i++];
    } else {
      reordered[k] = line[j--];
    }

    k++;
    state = !state;
  }

  return reordered;
}

File: 2018/solutions/prob05.java
import java.io.*;

public class prob05 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader
                = new BufferedReader(new InputStreamReader(System.in));

        String line = null;
        while ((line = reader.readLine()) != null) {
            String word = line;
            System.out.printf("%s%n", fold(word));
        }
    }

    public static String fold(String line) {

        boolean state = true;
        int lineLength = line.length();
        int k = 0, i = 0, j = lineLength - 1;
        String reordered = "";

        while (i <= j) {
            if (state) {
                reordered += line.charAt(i++);
            } else {
                reordered += line.charAt(j--);
            }

            k++;
            state = !state;
        }

        return reordered;
    }
}
